src/main/java/de/fraunhofer/aisec/cpg/graph/NamespaceDeclaration.java

```diff
@@ -26,4 +26,104 @@
 
 package de.fraunhofer.aisec.cpg.graph;
 
-public class NamespaceDeclaration extends TranslationUnitDeclaration {}
+import de.fraunhofer.aisec.cpg.helpers.NodeComparator;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.checkerframework.checker.nullness.qual.NonNull;
+
+/**
+ * Declares the scope of a namespace and appends its own name to the current namespace-ürefix to
+ * form a new namespace prefix. While RecordDeclarations in C++ and Java have their own namespace,
+ * namespace declarations can be declared multiple times. At the beginning of a Java-file, a
+ * namespace declaration is used to represent the package name as namespace. In its explicit
+ * appearance a namespace declaration can contain {@link FunctionDeclaration}, {@link
+ * FieldDeclaration} and {@link RecordDeclaration} similar to a {@link RecordDeclaration} and the
+ * semantical difference between NamespaceDeclaration and {@link RecordDeclaration} lies in the
+ * non-instantiabillity of a namespace.
+ */
+public class NamespaceDeclaration extends Declaration {
+
+  /** Edges to a {@link FieldDeclaration} defined in a namespace, these fields are static. */
+  @SubGraph("AST")
+  private List<FieldDeclaration> fields = new ArrayList<>();
```

Comment:
Why does this argument not hold for records? You could argue that a Record should only have a declarations field. I don't like the utility function because those are not available to graph queries later. Having edges that encode the specific type of declaration is not a problem IMO. However, I see the argument that ProblemDeclarations, which will be present wherever we have declarations meaning we have to switch to default declarations for every Node that contains declarations including RecordDeclarations.

Reply:
This is probably something that we can improve once we have edge properties. Then we can annotate the declarations edges with the correct type (at least I guess that this will be possible :man_shrugging:)

Reply:
I am not sure if this will be at all possible in Codyze because of OverflowDB. Afaik edges only exist "virtually" in overflowdb and I am not sure if they support properties on edges.

Reply:
Edges do have properties there, if you look at the following code:

https://github.com/Fraunhofer-AISEC/codyze/blob/51ce074e262009234cf82a5e9aa72acc9b909f7f/src/main/java/de/fraunhofer/aisec/crymlin/connectors/db/OverflowDatabase.java#L968-L1000

At the moment this is done via an `@EdgeProperty` annotation that is present in the CPG, but I guess we could use different mechanisms with more fine-grained control, e.g. some kind of wrapper like `List<AnnotatedNode>` where an `AnnotatedNode` consists of a target node and a list of edge properties, which is then unpacked in the codyze OGM. At least OverflowDB edges do have the possibility to assign properties to each one